fem/intrules.hpp

```diff
@@ -257,10 +263,112 @@ class IntegrationRule : public Array<IntegrationPoint>
        a call like this: `IntPoint(i).weight`. */
    const Array<double> &GetWeights() const;
 
+   /// Return an integration rule for KnotVector @a kv, defined by applying this
+   /// rule on each knot interval.
+   IntegrationRule* ApplyToKnotIntervals(KnotVector const& kv) const;
+
    /// Destroys an IntegrationRule object
    ~IntegrationRule() { }
 };
 
+/// Class for defining different integration rules on each NURBS patch.
+class NURBSPatchRule
+{
+public:
+   /// Construct a rule for each patch, using SetPatchRule.
+   NURBSPatchRule(const int numPatches, const int dim_) :
+      patchRule(numPatches), patchRules1D(numPatches, dim_),
+      npatches(numPatches), dim(dim_)
+   {
+      patchRule = nullptr;
+   }
+
+   /// Construct a tensor product of 1D rules, to be applied to all patches.
+   NURBSPatchRule(IntegrationRule *irx, IntegrationRule *iry,
+                  IntegrationRule *irz = nullptr);
+
+   /// Returns a rule for the element.
+   IntegrationRule &GetElementRule(const int elem, const int patch, int *ijk,
+                                   Array<const KnotVector*> const& kv) const;
+
+   /// Set the integration rule for the NURBS patch of the given index.
+   void SetPatchRule(const int patch, IntegrationRule *ir_patch)
+   {
+      patchRule[patch] = ir_patch;
+   }
+
+   /// Add a rule to be used for individual elements. Returns the rule index.
+   std::size_t AddElementRule(IntegrationRule *ir_element)
+   {
+      elementRule.push_back(ir_element);
+      return elementRule.size() - 1;
+   }
+
+   /// Set the integration rule for the element of the given index. This rule is
+   /// used instead of the rule for the patch containing the element.
+   void SetElementRule(const std::size_t element,
+                       const std::size_t elementRuleIndex)
+   {
+      elementToRule[element] = elementRuleIndex;
+   }
+
+   /// Set 1D integration rules to be used as a tensor product rule on the patch
+   /// with index @a patch.
+   void SetPatchRules1D(const int patch,
+                        std::vector<const IntegrationRule*> & ir1D);
+
+   /// For tensor product rules defined on each patch by SetPatchRules1D,
+   /// return a pointer to the 1D rule in the specified @a dimension.
+   const IntegrationRule* GetPatchRule1D(const int patch,
+                                         const int dimension) const
+   {
+      return patchRules1D(patch, dimension);
+   }
+
+   /// For tensor product rules defined on each patch by SetPatchRules1D,
+   /// return the integration point with index (i,j,k).
+   void GetIntegrationPointFrom1D(const int patch, int i, int j, int k,
+                                  IntegrationPoint & ip);
+
+   /// Set pointToElem, used by GetPointElement.
+   void SetPointToElement(Mesh const& mesh);
```

Comment:
This method name and its comment do describe what the method does but without knowing what `pointToElem` contains and how it will be used this name might be too vague. I would suggest something more descriptive like `BuildPatchIJKToElementMap`. The name doesn't need to be that cumbersome but public methods should be somewhat descriptive. I would then include a reference to `GetPointElement` so that users are directed to the documentation provided there.

Reply:
I changed the name to `Finalize`, which is more understandable and generic.